### PR TITLE
Improve SuspendedTabletsIT reliability

### LINKED ISSUE #2121: NoSuchMethodException using HostRegexTableLoadBalancer
The runtime class loading of the `HostRegexTableLoadBalancer` is broken. It is likely this happened when it was refactored during the move to SPI. This exception can be seen in the manager log when running `SuspendedTabletsIT`, which seems to timeout often for me locally but the test can still pass even with this exception.
<pre>
2021-05-27T10:39:11,473 [balancer.TableLoadBalancer] WARN : Failed to load table balancer class org.apache.accumulo.core.spi.balancer.HostRegexTableLoadBalancer for table !0
java.lang.NoSuchMethodException: org.apache.accumulo.core.spi.balancer.HostRegexTableLoadBalancer.<init>(org.apache.accumulo.core.data.TableId)
        at java.lang.Class.getConstructor0(Class.java:3349) ~[?:?]
        at java.lang.Class.getConstructor(Class.java:2151) ~[?:?]
        at org.apache.accumulo.core.spi.balancer.TableLoadBalancer.constructNewBalancerForTable(TableLoadBalancer.java:58) ~[accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.core.spi.balancer.TableLoadBalancer.getBalancerForTable(TableLoadBalancer.java:91) [accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.core.spi.balancer.TableLoadBalancer.getAssignments(TableLoadBalancer.java:115) [accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.manager.Manager.getAssignments(Manager.java:1735) [accumulo-manager-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.manager.TabletGroupWatcher.getAssignmentsFromBalancer(TabletGroupWatcher.java:880) [accumulo-manager-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.manager.TabletGroupWatcher.flushChanges(TabletGroupWatcher.java:911) [accumulo-manager-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.manager.TabletGroupWatcher.run(TabletGroupWatcher.java:318) [accumulo-manager-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
2021-05-27T10:39:11,474 [balancer.TableLoadBalancer] INFO : Using balancer org.apache.accumulo.core.spi.balancer.SimpleLoadBalancer for table !0
</pre>

### PR DESCRIPTION
**Pull request by brianloss**
The previous attempt to fix SuspendedTabletsIT attempted to use
HostRegexTableLoadBalancer as a per-table balancer. Unfortunately,
HostRegexTableLoadBalancer was not designed to be used that way but
instead was meant to replace TableLoadBalancer and use regular
expressions to divide the hosts into pools. Therefore, the balancer was
failing to initialize and wasn't actually doing any work.

* Performed minor refactoring on HostRegexTableLoadBalancer so that it
  can be extended to include the port when performing regex checks.
* Extend HostRegexTableLoadBalancer and configure it to use the port as
  well as host name for regex checks so that a single tserver on a host
  with multiple tservers running can be isolated and used for a table.
  In particular, this is used to send all metadata table tablets to a
  single tserver running in MiniAccumuloCluster.
* Start MiniAccumuloCluster with a single tserver so that we can use
  that information to configure the extended HostRegexTableLoadBalancer
  to send all metadata tablets to that first tablet server. With one
  tablet server running, we can also grab the process reference to it
  for when we later kill tablet servers (to ensure we don't kill the
  tablet server hosting the metadata table). Once this information is
  collected, then MiniAccumuloCluster is re-configured for 3 tablet
  servers and the additional tablet servers are launched.
* Fix the "Waiting on hosting and balance" conditions to wait for BOTH
  all tablets to be hosted and for those tablets to be hosted across all
  tablet servers (except the one reserved for the metadata table)
* Fix the "Waiting on suspended tablets" condition which was incorrectly
  using a conjunction when a disjunction was needed. The result was the
  loop could exit without any tablets actually reaching the suspended
  state.
* Refactor TabletLocations so we can choose to scan either the root or
  metadata table. This allows us to retrieve tablet locations for the
  metadata table.

Fixes #2121

### CODE REVIEW

**Comment by Manno15** [comment]
Tests are passing locally. I will take a closer look at the code now.

File: test/src/main/java/org/apache/accumulo/test/manager/SuspendedTabletsIT.java [hunk 1]
```diff
@@ -140,6 +193,7 @@ public void shutdownAndResumeTserver() throws Exception {
       }
 
       // remove servers with metadata on them from the list of servers to be shutdown
+      assertEquals("Expecting a single tServer in metadataServerSet", 1, metadataServerSet.size());
       tserverSet.removeAll(metadataServerSet);
 
       assertEquals("Expecting two tServers in shutdown-list", 2, tserverSet.size());
```
Comment (DomGarguilo) [inline test/src/main/java/org/apache/accumulo/test/manager/SuspendedTabletsIT.java:199]:
```suggestion
      assertEquals("Expecting " + (TSERVERS - 1) + " tServers in shutdown-list", TSERVERS - 1, tserverSet.size());
```

Comment (DomGarguilo) [inline test/src/main/java/org/apache/accumulo/test/manager/SuspendedTabletsIT.java:199]: Could add another assert around this line to ensure `tserverSet.size() >= count`

File: test/src/main/java/org/apache/accumulo/test/manager/SuspendedTabletsIT.java [hunk 2]
```diff
@@ -285,6 +341,48 @@ public static void cleanup() {
...
   }
 
+  /**
+   * A version of {@link HostRegexTableLoadBalancer} that includes the tablet server port in
+   * addition to the host name when checking regular expressions. This is useful for testing when
+   * multiple tablet servers are running on the same host and one wishes to make pools from the
+   * tablet servers on that host.
+   */
+  public static class HostAndPortRegexTableLoadBalancer extends HostRegexTableLoadBalancer {
+    private static Logger LOG =
```
Comment (DomGarguilo) [inline test/src/main/java/org/apache/accumulo/test/manager/SuspendedTabletsIT.java:351]:
```suggestion
    private static final Logger LOG =
```

**Review by DomGarguilo — COMMENTED** [review]
These look like some solid improvements. Tests pass locally. Just a few minor suggestions/discussion points and everything else looks good to me.

File: test/src/main/java/org/apache/accumulo/test/manager/SuspendedTabletsIT.java (hunk 1 above)
Comment (brianloss) [reply test/src/main/java/org/apache/accumulo/test/manager/SuspendedTabletsIT.java:199]: This isn't a line I changed, but I did notice that the value 2 is hard-coded a bunch of places in the test. If I'm going to fix one, then they should probably all be fixed.

File: test/src/main/java/org/apache/accumulo/test/manager/SuspendedTabletsIT.java [hunk 3]
```diff
@@ -91,21 +100,65 @@ public void configure(MiniAccumuloConfigImpl cfg, Configuration fsConf) {
...
   public void crashAndResumeTserver() throws Exception {
     // Run the test body. When we get to the point where we need a tserver to go away, get rid of it
     // via crashing
     suspensionTestBody((ctx, locs, count) -> {
-      List<ProcessReference> procs =
-          new ArrayList<>(getCluster().getProcesses().get(ServerType.TABLET_SERVER));
+      // Exclude the tablet server hosting the metadata table from the list and only
+      // kill tablet servers that are not hosting the metadata table.
+      List<ProcessReference> procs = getCluster().getProcesses().get(ServerType.TABLET_SERVER)
+          .stream().filter(p -> !metadataTserverProcess.equals(p)).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (brianloss) [reply test/src/main/java/org/apache/accumulo/test/manager/SuspendedTabletsIT.java:162]: Done

**Review by DomGarguilo — APPROVED** [review]

**Review by Manno15 — APPROVED** [review]

**Pull request merged**